src/core/lifetime.d

```diff
@@ -1326,6 +1332,44 @@ void copyEmplace(S, T)(ref S source, ref T target) @system
     assert(target.x == 42);
 }
 
+// preserve shared-ness
+@system pure nothrow unittest
+{
+    auto s = new Object();
+    auto ss = new shared Object();
+
+    Object t;
+    shared Object st;
+
+    copyEmplace(s, t);
+    assert(t is s);
+
+    copyEmplace(ss, st);
+    assert(st is ss);
+
+    static assert(!__traits(compiles, copyEmplace(s, st)));
+    static assert(!__traits(compiles, copyEmplace(ss, t)));
+}
+
+// don't violate immutability for reference types
+@system pure nothrow unittest
+{
+    auto s = new Object();
+    auto si = new immutable Object();
+
+    Object t;
+    immutable Object ti;
+
+    copyEmplace(s, t);
+    assert(t is s);
+
+    copyEmplace(si, ti);
+    assert(ti is si);
```

Comment:
Any idea why this assertion above would fail for 32-bit x86 debug (release works)? DMD codegen bug?

Reply:
Disabled for DMD x86 Posix now as workaround.